sdks/python/apache_beam/io/gcp/bigquery_tools.py

```diff
@@ -310,14 +312,19 @@ def _get_query_results(self, project_id, job_id,
   @retry.with_exponential_backoff(
       num_retries=MAX_RETRIES,
       retry_filter=retry.retry_on_server_errors_filter)
```

Comment:
https://github.com/apache/beam/blob/212a2ab95f87df0d2670873eeafd7e96c993c732/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java#L739

Reply:
Thanks for the feedback. I looked into the [error codes](https://cloud.google.com/bigquery/troubleshooting-errors) for BQ results, and you are right that the 403 error is often retriable (quota, rate issues). I can create a new filter that includes server errors, timeouts and 403 errors. What do you think about that? I believe HttpError would be the catchall for IOException in this case, as the library uses apitools: https://github.com/google/apitools/blob/master/apitools/base/py/http_wrapper.py#L313-L403

Reply:
I think it would be nice to have `always_retry` for inserting API calls just like https://github.com/apache/beam/blob/1fb4740775fe58fa16a146834d16082155ae2bab/sdks/python/apache_beam/utils/retry.py#L128 as the same behavior across the SDKs in different languages might give better usability (and less surprise).

Reply:
This insert method could be used on batch pipelines, and users may want to have the pipeline fail instead of having infinite retries, no?

The transform already takes a retry_strategy parameter that allows them to specify RETRY_ALWAYS. The only issue with these retries is that they don't have the exponential backoff algorithm. Would you think that adding exponential backoff there should be enough?

Reply:
> This insert method could be used on batch pipelines, and users may want to have the pipeline fail instead of having infinite retries, no?

You're right.

> 
> The transform already takes a retry_strategy parameter that allows them to specify RETRY_ALWAYS. The only issue with these retries is that they don't have the exponential backoff algorithm. Would you think that adding exponential backoff there should be enough?

Make sense. I think exponential backoff + (Optional) RETRY_ALWAYS should be enough.